Support decoupled scraping in cloudwatch_exporter integration.

dtrejod opened the pull request:
#### PR Description

Decouple the collection of cloudwatch metrics from the actual scrape on the `/metrics` endpoint. Decouple scraping is the default behavior in the upstream [YACE](https://github.com/nerdswords/yet-another-cloudwatch-exporter/blob/master/README.md) repo. 

In the grafana-agent implementation of decouple-scraping, I made the feature opt-in. The previous behavior is preserved unless users add to the `cloudwatch_exporter` integration the following config:
```
integrations:
    cloudwatch_exporter: 
        decoupled_scraping:
            enabled: true
```

With the above configuration decouple-scraping will be enabled. Scrapes from cloudwatch will occur every 5m by default unless overriden by `integrations.cloudwatch_exporter.decoupled_scraping.scraping_interval`.

#### Which issue(s) this PR fixes

https://github.com/grafana/agent/issues/4251

#### Notes to the Reviewer

The existing integration didn't have any tests. I was able to validate the PR here runs in a dev environment I have. Happy to add tests if given guidance on what those would look like in this repo. 

Also not sure what is meant by `Documentation added`? I think docs are auto-generated from docstrings? If so there is documentation for the new feature. 

#### PR Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] CHANGELOG.md updated
- [ ] Documentation added
- [ ] Tests updated
- [ ] Config converters updated

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/agent?pullRequest=4914) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/agent?pullRequest=4914) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/agent?pullRequest=4914) it.</sub>

dtrejod commented on pkg/integrations/cloudwatch_exporter/config.go:
```diff
@@ -17,10 +17,11 @@ import (
+	cloudWatchConcurrency            = 5
+	tagConcurrency                   = 5
+	labelsSnakeCase                  = false
+	defaultDecoupledScrapingInterval = time.Minute * 5
```
5m is the default scrape interval in YACE:
https://github.com/nerdswords/yet-another-cloudwatch-exporter/blob/master/cmd/yace/main.go#L133

marctc commented:
> Also not sure what is meant by `Documentation added`? I think docs are auto-generated from docstrings? If so there is documentation for the new feature.

`Documentation added` checkbox it's a reminder in the case that PR needs to included changes in the docs. In this case, it's applicable as you added a new field, which would be needed to be documented.

I'd defer the review to @thepalbi, but this PR needs to add the same fields for the flow component. Could you do that as well? thanks!

mattdurham commented:
@dtrejod I cannot add this since its a fork but can you add this to the doc in cloudwatch-exporter-config?

```
# Instead of retrieving metrics on request, decouple scraping retrieves the metrics on a schedule and returns the cached metrics.
  decoupled_scraping:
    # Enable decoupled scraping.
    [enabled: <boolean> | default = false ]

    # How often to scrape the endpoint.
    [scrape_interval: <duration> | default = "5m"]
```

dtrejod commented:
> Documentation added checkbox it's a reminder in the case that PR needs to included changes in the docs. In this case, it's applicable as you added a new field, which would be needed to be documented.
> I cannot add this since its a fork but can you add this to the doc in cloudwatch-exporter-config?

Thanks both. I missed the "docs" pkg when looking at the root dir structure of this repo. Added docs to both static and flow integrations. 

Also updated the flow integration to also have the `decoupled scraping` feature. 

> Can I ask what's the driver for preferring the scrape to be asynchronous (like yace does natively) vs. during the /metrics call?

The main driver is to fix `write i/o timeout` errors we are seeing in some of our largest AWS cloud accounts. Today we get errors such as:
```
GET /integrations/cloudwatch_exporter/metrics 53.47008076s, error: write tcp 127.0.0.1:5080->127.0.0.1:55386: i/o timeout
```

When tried a few idea to work around the issue such as bumping the `scrape_interval`/`scrape_timeout` for the component integration to `5m`. However this causes the cloudwatch metrics collected by grafana-agent to be sampled at `5m` vs `1m` for all other integrations. The differing sample frequency futher causes issues when viewing these metrics in Grafana since with a `5m` scrape interval we couldn't use `$rate_interval` and other built-in Grafana variables with our `rate()` funcs. 

The decoupled scraping solves the io timeouts while also allowing us to keep the sample frequency at `1m`.

clayton-cornell commented on docs/sources/flow/reference/components/prometheus.exporter.cloudwatch.md:
```diff
@@ -323,6 +326,21 @@ is exported to CloudWatch.
+| Name              | Type     | Description                                                             | Default | Required |
+|-------------------|----------|-------------------------------------------------------------------------|---------|----------|
+| `enabled`         | `bool`   | Controls whether the decoupled scraping featured is enabled             | false   | no       |
+| `scrape_interval` | `string` | Controls how frequently to asynchronously gather new cloudwatch metrics | 5m      | no       |
```
```suggestion
| `scrape_interval` | `string` | Controls how frequently to asynchronously gather new CloudWatch metrics | 5m      | no       |
```

clayton-cornell commented on docs/sources/static/configuration/integrations/cloudwatch-exporter-config.md:
```diff
@@ -132,6 +132,16 @@ Configuration reference:
+    # Enable decoupled scraping.
+    [enabled: <boolean> | default = false ]
+
+    # How often to scrape for AWS metrics.
```
Which is it? AWS metrics? or CloudWatch? The rest of the docs refer to it as CloudWatch metrics.

dtrejod replied:
Sorry this was written before the flow docs below. I think CloudWatch is more consistent, so I'll update this.

clayton-cornell reviewed (commented):
Some comments and questions for docs.

thepalbi commented:
> > Documentation added checkbox it's a reminder in the case that PR needs to included changes in the docs. In this case, it's applicable as you added a new field, which would be needed to be documented.
> > I cannot add this since its a fork but can you add this to the doc in cloudwatch-exporter-config?
> 
> Thanks both. I missed the "docs" pkg when looking at the root dir structure of this repo. Added docs to both static and flow integrations.
> 
> Also updated the flow integration to also have the `decoupled scraping` feature.
> 
> > Can I ask what's the driver for preferring the scrape to be asynchronous (like yace does natively) vs. during the /metrics call?
> 
> The main driver is to fix `write i/o timeout` errors we are seeing in some of our largest AWS cloud accounts. Today we get errors such as:
> 
> ```
> GET /integrations/cloudwatch_exporter/metrics 53.47008076s, error: write tcp 127.0.0.1:5080->127.0.0.1:55386: i/o timeout
> ```
> 
> When tried a few idea to work around the issue such as bumping the `scrape_interval`/`scrape_timeout` for the component integration to `5m`. However this causes the cloudwatch metrics collected by grafana-agent to be sampled at `5m` vs `1m` for all other integrations. The differing sample frequency futher causes issues when viewing these metrics in Grafana since with a `5m` scrape interval we couldn't use `$rate_interval` and other built-in Grafana variables with our `rate()` funcs.
> 
> The decoupled scraping solves the io timeouts while also allowing us to keep the sample frequency at `1m`.

The setup you are working with, is it grafana agent configured with the cloudwatch integration, and a destination prometheus to write on; or are you using the integration, and making prometheus scrape the `/integrations/cloudwatch_exporter/metrics` endpoint?

As shown in the docs, you can configure the `scrape_interval` and `scrape_timeout` just for the CloudWatch integrations. In that case, have you tried keeping the `1m` global scrape interval, and just bumping the `scrape_timeout` on the CloudWatch integration to sth that allows the scrape to run? 

For the record, we've seen internally in YACE (which is basically what the agent uses underneath) that for big big jobs (like you mention big AWS accounts) some scrape times are in the order of 1/2 mins.

dtrejod commented on docs/sources/flow/reference/components/prometheus.exporter.cloudwatch.md:
```diff
@@ -323,6 +326,21 @@ is exported to CloudWatch.
+
+The decoupled scraping feature is useful to avoid against abuse of API requests that can caused extra billing in an AWS
+account. It is also useful to avoid scrape timeouts when running in very large AWS accounts where the number of
+cloudwatch metrics being gathered can not completed before the component scrape timeout.
```
It's unclear what the exact number of AWS resources causes the scrape timeout to be hit. It's also not just tied to the size of the AWS account, but also what the user has configured to be scraped via this cloudwatch_exporter integration. 

I'm inclined to leave it purposefully vague by just saying "high"?

thepalbi commented:
Not against this change, but just want to make sure if you tried this other suggestion. Have you tried:
- increasing the `scrape_timeout` on the CW integration, to something larger that `1m`, that it allows your big scrapes to complete
- leaving the `scrape_interval` at `1m` for every integration

In that case you'll keep the `1m` scrapes everywhere, but allowing more time for the big ones, so that they are able to finish. This won't give you the caching mechanism async does, which is useful if you are scraping for example ec2 metrics that are produced every 5m, but still need the 1m scrape interval to keep everything aligned, but might work.

mattdurham reviewed (approved):
Thank you for the PR, LGTM after the feedback

dtrejod commented:
> increasing the scrape_timeout on the CW integration, to something larger that 1m, that it allows your big scrapes to complete leaving the scrape_interval at 1m for every integration

Yes we tried this and still saw consistent timeouts. We weren't able to diagnose the exact reason why these cloudwatch timeouts would happen even with very large scrape_timeout config. The below full structured log line shows you some more information as to the weirdness we saw. You can see in the headers that clearly show our scrape-timoeut was set to 2min yet we still timed out at 1m.. 

```
{"caller":"gokit.go:79","level":"warn","msg":"GET /integrations/cloudwatch_exporter/metrics 53.47008076s, error: write tcp 127.0.0.1:5080->127.0.0.1:55386: i/o timeout ws: false; Accept: application/openmetrics-text;version=1.0.0,application/openmetrics-text;version=0.0.1;q=0.75,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: GrafanaAgent/v0.35.1; X-Prometheus-Scrape-Timeout-Seconds: 120; ","ts":"2023-08-17T15:25:07.33575929Z"}
```

Overall this decoupled-scraping feature is definitely a solution that is working for us and one we'll start using right away. 

Thanks for the quick reviews!

thepalbi commented:
Thanks to you for the work @dtrejod !
